### ISSUE TITLE
Exclude lists aren't deep merged

### ISSUE THREAD

**Issue by betesh** [bbatsov/rubocop#4477]
When merging 2 configurations, `Exclude` lists in the first one get overwritten by `Exclude` lists for the same cop in the second one.

--------

## Expected behavior

In the example below, I expect both a.rb and b.rb to be excluded, and only the Gemfile to be inspected.

## Actual behavior
The file listed in the inherited rubocop configuration (a.rb) is inspected.

```
$ rubocop 
Inspecting 2 files
.C

Offenses:

a.rb:1:1: C: Missing top-level class documentation comment.
class A
^^^^^

2 files inspected, 1 offense detected
```

## Steps to reproduce the problem

```
$ cat a.rb 
class A
  def my_method; end
end

$ cat b.rb 
class B
  def my_method; end
end

$ cat Gemfile
source 'https://rubygems.org'
gem 'rubocop', git: 'https://github.com/bbatsov/rubocop.git'

$ cat .rubocop.yml 
inherit_from: .rubocop_1.yml

AllCops:
  Exclude:
    - b.rb

$ cat .rubocop_1.yml 
AllCops:
  Exclude:
    - a.rb

```

## RuboCop version

```
$ rubocop -V
0.49.1 (using Parser 2.4.0.0, running on ruby 2.3.4 x86_64-linux)
```

**Comment by Drenmi** [comment]
This behaviour is intentional. Without it, there would be no way of removing items added from a previous configuration file (e.g. `.default.yml`). 🙂 

I was thinking earlier if there's a keyword we could introduce to force merging, but I couldn't come up with something that's intuitive and concise enough to make a suggestion.

**Comment by jonas054** [comment]
I have a feeling we've talked about it before, but decided against it. I can't find any record of that discussion though so it might be a false memory. I was thinking about something like
```yaml
AllCops:
  Exclude:
    - inherit
    - a.rb
```

**Comment by betesh** [comment]
>  If arrays were merged, there would be no way to remove elements through overriding them in local configuration.

We cetainly need a way to remove elements, but IMO we're doing it in a way that violates the [Principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

Perhaps we should take the opposite approach: Arrays should be merged by default (least surprising behavior), but give us a way to exclude items already added.  Perhaps this:

```
AllCops:
  Exclude:
    - b.rb # Equivalent to "b.rb: True"
    - a.rb: False
```

This way, every Array is effectively a Hash, where if you just list the key, the implicit value is `True`.

If this sounds reasonable, then the next step is to come up with an upgrade path because this is a breaking change.

**Comment by Drenmi** [comment]
> IMO we're doing it in a way that violates the Principle of least surprise.

It is surprising if you only consider the special case of multiple configuration files. In the base case, the opposite is confusing. Files being excluded even though they are not in the configuration file exclude list, because it was merged with the `.default.yml` which is hidden away in the gem, would almost certainly baffle a lot of casual users.

> This way, every Array is effectively a Hash [...]

This definitely sounds like a violation of TPoLS to me. 😉 

Another option could be to add an additional pragma. E.g.:

```yaml
inherit_from: ...
inherit_mode: merge
```

which defaults to `override`. This would make it backwards compatible, opt-in, and explicit.

That said, @bbatsov has expressed regrets for supporting multiple configuration files before, and I suspect digging the hole even deeper is not a direction he'd like to go, but he'd have to speak for himself, of course. 😉

**Comment by betesh** [comment]
>> This way, every Array is effectively a Hash [...]

> This definitely sounds like a violation of TPoLS to me. :wink:

Not really...it's behind the scenes.  If you never need to remove anything from a list, the existing syntax would continue to work for you and you wouldn't know anything changed.

With `inherit_mode`, we'd have a few design issues to address:

* If A inherits from B with mode `override` and B inherits from C with mode `merge`, what happens to the Array elements from C?  Do they get overridden by A because from A's standpoint, they are part of B?  Or do they get merged into A because A is only interested in overriding the lists in B, not lists from other configuration?

* What if some lists in A need to be merged with their corresponding lists in B, and others need to overwrite?  (My suggestion above allows setting the behavior per Array element, whereas this suggestion allows setting it per config file)

**Comment by Drenmi** [comment]
> If A inherits from B with mode override and B inherits from C with mode merge, what happens to the Array elements from C?

Using `override` would override any inherited settings. Anything else would be very surprising, not backwards compatible, and re-open the issue that some array elements can't be cleared.

> What if some lists in A need to be merged with their corresponding lists in B, and others need to overwrite?

I think supporting this level of complexity would actually be a disservice to the users. We'd essentially spend effort on introducing ways for them to over-complicate their configuration files.

**Comment by jhottenstein** [comment]
Hello,

As a user, I find the current behavior difficult when I am trying to move items from my todo to the regular yaml.

My use case is:
- I have lots of items in my .rubocop_todo.yml
- I try to improve a single file.  
- I remove most of the items from the .rubocop_todo.yml
- I move one item to the .rubocop.yml with an exclude because I don't want the cop to run on that file/directory
- Now rubocop flags all items for that cop in .rubocop_todo.yml

I don't necessarily need configurations to be merged but I would like some way to support that workflow.

Thanks for a great project!

**Comment by jonas054** [comment]
> Now rubocop flags all items for that cop in .rubocop_todo.yml

I don't quite follow. Can you show us the configuration for the cop in question, in `.rubocop.yml` and in `.rubocop_todo.yml`, plus the output from RuboCop?

**Comment by jhottenstein** [comment]
Sorry for the delay:
With the following configuration:

.rubocop_todo.yml
```
# Offense count: 22
RSpec/MultipleExpectations:
  Exclude:
    - 'spec/models/expense_spec.rb'
    - 'spec/models/group_spec.rb'
    - 'spec/models/notification_spec.rb'
    - 'spec/requests/api_spec.rb'
    - 'spec/requests/get_main_data_spec.rb'
    - 'spec/requests/invite_spec.rb'
```
.rubocop.yml
```
inherit_from: .rubocop_todo.yml

RSpec/MultipleExpectations:
  Max: 5
```

I'll get something like this: 
```
$ bin/rubocop
Inspecting 357 files
....................................................................................................................................................................................................................................................C................................................................................................................

Offenses:

spec/requests/group_invite_spec.rb:27:7: C: RSpec/MultipleExpectations: Example has too many expectations [8/5].
      it 'joins the group' do
      ^^^^^^^^^^^^^^^^^^^^

357 files inspected, 1 offense detected
```

Now if i want to exclude that file, I would add an exclude to .rubocop.yml
.rubocop.yml
```
inherit_from: .rubocop_todo.yml

RSpec/MultipleExpectations:
  Max: 5
  Exclude:
    - spec/requests/group_invite_spec.rb
```
But then i get this from rubocop
```
.
.
.
spec/requests/get_main_data_spec.rb:12:3: C: RSpec/MultipleExpectations: Example has too many expectations [12/5].
  it 'returns all data as expected' do
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

357 files inspected, 23 offenses detected
```
